refactor(popup): Convert Popup to a compound component (#1065)
Workday/canvas-kit: modules/react/popup/stories/Popup.stories.mdx

import {Meta} from '@storybook/addon-docs/blocks';

import {Popup} from '@workday/canvas-kit-react/popup';
import {Specifications} from '@workday/canvas-kit-docs';

import {Simple} from './examples/Simple';
import {InitialFocus} from './examples/InitialFocus';
import {OpenWithoutModel} from './examples/OpenWithoutModel';
import {ManualControl} from './examples/ManualControl';
import {MultiplePopups} from './examples/MultiplePopups';
import {NestedPopups} from './examples/NestedPopups';
import {RTL} from './examples/RTL';
import {UsePopup} from './examples/UsePopup';
import {PopupModelConfigComponent, PopupStateComponent, PopupEventsComponent} from './PopupModel';

<Meta title="Components/Popups/Popup/React" component={Popup} />

# Canvas Kit Popups

A "popup" is a classification for a type of stacked UI element that appears "on top" of statically
positioned content. Tooltips, Modals, Dropdown menus, etc are all examples of "popups". Canvas Kit
has a "stack manager" system for managing these stacked UIs. Different types of popups have
different requirements of behavior for UX and accessibility - we can call them behaviors,

> capabilities, or traits. Canvas Kit comes with a number of [behaviors](#hooks) in the form of React
> Hooks.

Reviewer: Nit: If the link is to a hooks section, the hooks reference should be included in the link

```suggestion
capabilities, or traits. Canvas Kit comes with a number of [behaviors in the form of React
Hooks](#hooks).
```

Author: What about "behavioral hooks"?

You should use the most semantic component for your use-case before using `Popup` directly, like

> `Modal`, which already has the correct behaviors built-in. If no component already exists, you can

Reviewer: ```suggestion
`Modal`, which already has the correct behaviors built-in. If no component already exists that matches your use case, you can
```

> use `Popup` and use our [hooks](#hooks). The `Popup` component comes with a `Popup.Popper` component

Reviewer: ```suggestion
use `Popup` and use our [hooks](#hooks). The `Popup` component comes with a `Popup.Popper` sub-component
```

Author: "sub-component" or "subcomponent"?

that positions a popup using [PopperJS](https://popper.js.org/) that registers a popup with the
`PopupStack` automatically and sets the popup model's `placement` property. `Popup.Popper` component
and hooks work with the stack management system for correct rendering and accessibility behavior. If
you cannot use `Popup.Popper`, use the [usePopupStack](#usepoupstack) hook to properly register and
deregister the popup at the correct time. If you cannot use our hooks, consider upgrading your

> component to use Hooks. If you cannot do that, you'll have to look up the `PopupStack` package for

Reviewer: > `PopupStack` package

Link?

Author: I think we should link between documents, but we don't have a way to do that well yet. Do we link to other stories? If yes, how do we know when those links break?

the direct API and have a look at the source code for our hooks into the `PopupStack` API.

This package comes with everything you need to build Popup UIs.

## Installation

```sh
yarn add @workday/canvas-kit-react
```

## Usage

> The `Popup` component is a low-level compound component that is used to build popup UIs that are not

Reviewer: ```suggestion
The `Popup` component is a generic catch-all compound component that is used to build popup UIs that are not
```

I don't know if I'd call it "low-level"

Author: I like "generic", but not "catch-all"

already covered by Canvas Kit.

### Basic

> The Popup has no pre-defined behaviors already registered, therefore the `usePopupModel` must always

Reviewer: > The Popup has no pre-defined behaviors already registered

Not really sure what you're trying to say with this.

Author: Behaviors like closing on escape key, closing when clicking outside, focus management, etc. I realize now that "behavior hook" is overloaded with compound components. "Popup behaviors" existed before "behavior hooks" were defined in our compound component doc...

Hooks like `useCloseOnEscape` are still behaviors, but they aren't directly tied to subcomponents like the compound component doc lays out. `useCloseOnEscape` operates only on the model and not on a direct element. I guess technically they operate on the `stack` element which isn't owned by React, but instead owned directly by the `PopupStack` and storied in the `PopupModel`.

Reviewer: Ah - I see. I think it was the term "registered" that confused me. Maybe just something like:
```suggestion
The Popup has no pre-defined behaviors built in, therefore the `usePopupModel` must always
```

be used to create a new `model`. This `model` is then used by all behavior hooks to apply additional
popup behaviors to the compound component group. The following example creates a typical popup
around a target element and adds `useCloseOnOutsideClick`, `useCloseOnEscape`, `useInitialFocus`,
and `useReturnFocus` behaviors. You can read through the [hooks](#hooks) section to learn about all
the popup behaviors. For accessibility, these behaviors should be included most of the time.

<ExampleCodeBlock code={Simple} />

### Without a model

This example shows how a `Popup.Card` can be used outside a `Popup` component, but accessibility is
lost. If you use the Popup this way, you'll have to manually add a `aria-labelledby` to the

> `Popup.Card` component and an `id` to the `Popup.Heading` component.

Reviewer: When/why would you want/need to do this?

Author: I don't think people should do this, but it is the easiest way to upgrade. I'm not sure where this type of transition documentation should be, but it will be questions we'll commonly get. I wouldn't want v5 adoption to be blocked by Popups being too difficult to migrate to for some teams.

Reviewer: Maybe we put that in the migration guide rather than in our storybook docs then?

> **Note:** This example shows an inaccessible open card for demonstration purposes.

<ExampleCodeBlock code={OpenWithoutModel} />

### Manually passing the model

Each `Popup` sub-component optionally takes a `model`. Sometimes you may want to render an open
Popup because visibility is controlled elsewhere. Behaviors require a model, so you can initialize
the model visibility set to `true`. This may be required for legacy uses or where the entire `Popup`
component API is undesired. This example has the same functionality as the [Basic](#basic) example,
but breaks up control of a Popup in a way that is common to the way Popups are done in v4. This is
an intermediate way to describe Popups with a more incremental upgrade path from version 4.

<ExampleCodeBlock code={ManualControl} />

### Initial focus

> If you want focus to move to a specific element, set the `initialFocusRef` of the model. Check with

Reviewer: ```suggestion
If you want focus to move to a specific element when the modal is opened, set the `initialFocusRef` of the model. Check with
```

accessibility before doing this. The following example sets the focus on the "OK" button with an
`aria-describedby` pointing to the model's `id` state so screen readers properly announce the

> message of the popup when focus is changed to the button.

Reviewer: ```suggestion
message of the popup when focus is changed to the button. By default, focus will be placed on the close button when the modal is opened.
```

> **Note:** This example shows an inaccessible open card for demonstration purposes.

<ExampleCodeBlock code={RTL} />

## Components

## Popup

This component is a container component that has no semantic element. It provides a React Context
model for all Popup sub-components. If you manually pass a `model` to all sub-components, this
container component isn't needed.

<ArgsTable of={Popup} />

## Popup.Target

A `Popup.Target` is any element that is meant to show the Popup. The default component rendered by
this component is a `Button` element. You can override this by passing the desired component via
`as`. Many examples above use `as={DeleteButton}`. If you want to render an `IconButton` instead, do
`as={IconButton}`. This element merges all props provided by the
[usePopupTargetButton](#usepopuptargetbutton) hook which would be used in place of this component if
desired.

```tsx
const model = usePopupModel();

// using this component
<Popup.Target>Show Popup</PopupTarget>

// using props instead
const popupTargetButtonProps = usePopupTargetButton(model);
<Button {...popupTargetButtonProps}>Show Popup</Button>
```

## Popup.Popper

> A `Popup.Popper` is a [Popper](#popper) component that is hooked up to the `PopupModel`

Reviewer: There's no `#popper` heading

automatically.

## Popup Model

The PopupModel contains all state and events needed for Popup behaviors and components. It composes
a `DisclosureModel` for showing and hiding content.

### Popup Model State

<ArgsTable of={PopupStateComponent} />

### Popup Model Events

<ArgsTable of={PopupEventsComponent} />

### Popup Model Config

<ArgsTable of={PopupModelConfigComponent} />

## Hooks

### usePopupStack

```ts
const stackRef = usePopupStack(forwardRef?: React.RefObject<HTMLElement>): React.RefObject<HTMLDivElement>
```

**Note:** If you're using `Popper`, you do not need to use this hook directly.

This hook will add the `stackRef` element to the Popup stack on mount and remove on unmount. If you
use `Popper`, the popper `stackRef` is automatically added/removed from the Popup stack. The Popup
stack is required for proper z-index values to ensure Popups are rendered correct. It is also
required for global listeners like click outside or escape key closing a popup. Without the Popup
stack, all popups will close rather than only the topmost one.

> If `forwardRef` is provided, it will be the same as `stackRef`. If `forwardRef` is not provided`,

Reviewer: ```suggestion
If `forwardRef` is provided, it will be the same as `stackRef`. If `forwardRef` is not provided,
```

this hook will create one and return it.

This hook should be used by all stacked UIs unless using the `Popper` component.

Example:

```tsx
const model = usePopupModel();
usePopupStack(model.state.stackRef, model.state.targetRef);

// add some popup functionality
useCloseOnOutsideClick(model);
useCloseOnEscape(model);

return (
  <>
    <button ref={model.state.targetRef}>Open Popup</button>
    {model.state.visible
      ? ReactDOM.createPortal(<div>Popup Contents</div>, model.state.stackRef.current)
      : null}
  </>
);
```

### usePopup

Convenience hook for common Popups used as non-modal dialogs. It provides props to mix into
composite parts of the Popup pattern.

> > Note: This is **deprecated**. Please upgrade to using `usePopupModel` and the `Popup` compound

Reviewer: I'm still confused about `usePopup` and `usePopupOld` if they're both deprecated haha. I still say let's just drop this. People can always use CK v5 with `@worday/canvas-kit-react-popup@4.x` if they need the old stuff temporarily.

Author: Hmm. That's fair. `usePopupOld` is a duplicate during testing.

`Modal`, which already has the correct behaviors built-in. If no component already exists, you can
use `Popup` and use our [hooks](#hooks). The `Popup` component comes with a `Popup.Popper` component
that positions a popup using [PopperJS](https://popper.js.org/) that registers a popup with the
`PopupStack` automatically and sets the popup model's `placement` property. `Popup.Popper` component
and hooks work with the stack management system for correct rendering and accessibility behavior. If
you cannot use `Popup.Popper`, use the [usePopupStack](#usepoupstack) hook to properly register and
deregister the popup at the correct time. If you cannot use our hooks, consider upgrading your
component to use Hooks. If you cannot do that, you'll have to look up the `PopupStack` package for
the direct API and have a look at the source code for our hooks into the `PopupStack` API.

This package comes with everything you need to build Popup UIs.

## Installation

```sh
yarn add @workday/canvas-kit-react
```

## Usage

The `Popup` component is a low-level compound component that is used to build popup UIs that are not
already covered by Canvas Kit.

> ### Basic

Reviewer: Nit: The story is called `Simple`, but the heading is `Basic`. Let's align on one.
